src/RBTree.mo

```diff
@@ -450,4 +412,249 @@ module {
+    x : X,
+    y : Y
+  )
+  : (Tree<X,Y>, ?Y) {
+    var y0 : ?Y = null;
+    func ins(tree : Tree<X,Y>) : Tree<X,Y> {
+      switch tree {
+        case (#leaf) {
+          #node(#R, #leaf, (x,?y), #leaf)
+        };
+        case (#node(#B, left, xy, right)) {
+          switch (compare (x, xy.0)) {
+            case (#less) {
+              lbalance(ins left, xy, right)
+            };
+            case (#greater) {
+              rbalance(left, xy, ins right)
+            };
+            case (#equal) {
+              y0 := xy.1;
+              #node(#B, left, (x,?y), right)
+            }
+          }
+        };
+        case (#node(#R, left, xy, right)) {
+          switch (compare (x, xy.0)) {
+            case (#less) {
+              #node(#R, ins left, xy, right)
+            };
+            case (#greater) {
+              #node(#R, left, xy, ins right)
+            };
+            case (#equal) {
+              y0 := xy.1;
+              #node(#R, left, (x,?y), right)
+            }
+          }
+        }
+      };
+    };
+    switch (ins tree) {
+      case (#node(#R, left, xy, right)) {
+        (#node(#B, left, xy, right), y0);
+      };
+      case other { (other, y0) };
+    };
+  };
+
+
+  func balLeft<X,Y>(left : Tree<X, Y>, xy : (X,?Y), right : Tree<X, Y>) : Tree<X,Y> {
+    switch (left, right) {
+      case (#node(#R, l1, xy1, r1), r) {
+        #node(
+          #R,
+          #node(#B, l1, xy1, r1),
+          xy,
+          r)
+      };
+      case (_, #node(#B, l2, xy2, r2)) {
+        rbalance(left, xy, #node(#R, l2, xy2, r2))
+      };
+      case (_, #node(#R, #node(#B, l2, xy2, r2), xy3, r3)) {
+        #node(#R,
+          #node(#B, left, xy, l2),
+          xy2,
+          rbalance(r2, xy3, redden r3))
+      };
+      case _ { Debug.trap "balLeft" };
+    }
+  };
+
+  func balRight<X,Y>(left : Tree<X, Y>, xy : (X,?Y), right : Tree<X, Y>) : Tree<X,Y> {
+    switch (left, right) {
+      case (l, #node(#R, l1, xy1, r1)) {
+        #node(#R,
+          l,
+          xy,
+          #node(#B, l1, xy1, r1))
+      };
+      case (#node(#B, l1, xy1, r1), r) {
+        lbalance(#node(#R, l1, xy1, r1), xy, r);
+      };
+      case (#node(#R, l1, xy1, #node(#B, l2, xy2, r2)), r3) {
+        #node(#R,
+          lbalance(redden l1, xy1, l2),
+          xy2,
+          #node(#B, r2, xy, r3))
+      };
+      case _ { Debug.trap "balRight" };
+    }
+  };
+
+  func append<X,Y>(left : Tree<X, Y>, right: Tree<X, Y>) : Tree<X, Y> {
+    switch (left, right) {
+      case (#leaf,  _) { right };
+      case (_,  #leaf) { left };
+      case (#node (#R, l1, xy1, r1),
+            #node (#R, l2, xy2, r2)) {
+        switch (append (r1, l2)) {
+          case (#node (#R, l3, xy3, r3)) {
+            #node(
+              #R,
+              #node(#R, l1, xy1, l3),
+              xy3,
+              #node(#R, r3, xy2, r2))
+          };
+          case r1l2 {
+            #node(#R, l1, xy1, #node(#R, r1l2, xy2, r2))
+          }
```

Comment:
I am also unsure here: Both parent and right node are red. For symmetry to the case at line 599, shouldn't there be a `balanceRight`?